src/effects/native/gatereffect.cpp

```diff
@@ -0,0 +1,236 @@
+    GaterGroupState::State state = pState->state;
+
+    // Use to keep track of the beginning of beats when playing with the rate
+    unsigned int beatPeriod = 1;
+    if (enableState == EffectProcessor::ENABLING) {        
+        // Start with the gate closed
+        state = GaterGroupState::IDLE;
+        gain = 0;
+
+        // Set initial time from current beat position, or 0 if no beat grid
+        if (groupFeatures.has_beat_length_sec && groupFeatures.has_beat_fraction) {
+            beatPeriod   = groupFeatures.beat_length_sec*sampleRate;
+            timePosition = groupFeatures.beat_fraction*beatPeriod;
+        } else {
+            timePosition = 0;   
+            beatPeriod   = sampleRate; // 1 second
+        }
+    }
+
+    // Adjust trigger period
+    unsigned int triggerPeriod;
+    double period = 1-rate;
+    if (groupFeatures.has_beat_length_sec) {
+        // triggerPeriod is a number of beats
+        if (m_pQuantizeParameter->toBool()) {
+            period = 1.0/divider;
+            
+            if (m_pTripletParameter->toBool()) {
+                period /= 3.0/4;
+            }
+        } else if (period < 1 / 8.0) {
+            period = 1 / 8.0;
+        }
+        triggerPeriod = period * groupFeatures.beat_length_sec * sampleRate;
+    } else {
+        // triggerPeriod is a number of seconds
+        period = std::max(period, 1 / 8.0);
+        triggerPeriod = period * sampleRate;
+    }
+
+    // triggerTime : offset for the gate trigger
+    unsigned int triggerTime;
+    if (m_pInvertParameter->toBool()) {
+        triggerTime = triggerPeriod/2;
+    } else {
+        triggerTime = 0;
+    }
+
+    // holdTime : number of samples spent in the Hold state
+    unsigned int holdTime;
+    double attackTime  = kMaxGain/attackInc;
+    double releaseTime = kMaxGain/releaseInc;
+    holdTime = std::max(0.0, floor(triggerPeriod*shape-attackTime-releaseTime));
+    
+    for (unsigned int i = 0; i < numSamples; i+=2) {
+         if (timePosition % triggerPeriod == triggerTime) {
+            state = GaterGroupState::ATTACK;
+            if (timePosition % beatPeriod == 1) {
+                timePosition = 0;
+            }
+        }
+        timePosition++;
+
+        // Gate state machine : Idle->Attack->Hold->Release->Idle
+        switch (state) {
+            // Idle : Gain = 0
+            case GaterGroupState::IDLE:
+                gain = 0;
+                break;
+
+            // Attack : Increase gain up to maxGain, then Hold
+            case GaterGroupState::ATTACK:
+                gain += attackInc;
+                if(gain >= kMaxGain) {
+                    state = GaterGroupState::HOLD;
+                    holdCounter = holdTime;
+                }
+                break;
+
+            // Hold : Hold maxGain for holdCounter samples
+            case GaterGroupState::HOLD:
+                gain = kMaxGain;
+                if(holdCounter == 0) {
```

Comment:
"if" is a keyword and not a function. Please insert a space between the keyword and the opening brace to improve readability. You can fix this when you need to edit the code anyway. Formatting the code according to syntactical rules should better be done with an automatic code formatter, that we don't have ;)